File: app/src/main/java/otang/network/util/TrafficUtils.java
package otang.network.util;

import android.content.Context;
import android.net.TrafficStats;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;

public class TrafficUtils {

	public static final long GB = 1000000000;
	public static final long MB = 1000000;
	public static final long KB = 1000;

	public static String getDownloadSpeed() {
		String speed, unit;
		Long bytesPrevious = TrafficStats.getTotalRxBytes();
		try {
			Thread.sleep(1000);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		Long bytesCurrent = TrafficStats.getTotalRxBytes();
		long networkSpeed = bytesCurrent - bytesPrevious;
		float downSpeed;
		if (networkSpeed >= GB) {
			downSpeed = networkSpeed / GB;
			unit = "GB/s";
		} else if (networkSpeed >= MB) {
			downSpeed = networkSpeed / MB;
			unit = "MB/s";
		} else {
			downSpeed = networkSpeed / KB;
			unit = "KB/s";
		}
		speed = String.format("%.1f", downSpeed);
		return speed + unit;
	}

	public static String getUploadSpeed() {
		String speed, unit;
		Long bytesPrevious = TrafficStats.getTotalTxBytes();
		try {
			Thread.sleep(1000);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		Long bytesCurrent = TrafficStats.getTotalTxBytes();
		long networkSpeed = bytesCurrent - bytesPrevious;
		float upSpeed;
		if (networkSpeed >= GB) {
			upSpeed = networkSpeed / GB;
			unit = "GB/s";
		} else if (networkSpeed >= MB) {
			upSpeed = networkSpeed / MB;
			unit = "MB/s";
		} else {
			upSpeed = networkSpeed / KB;
			unit = "KB/s";
		}
		speed = String.format("%.1f", upSpeed);
		return speed + unit;
	}

	public static String[] getNetworkSpeed() {
		String downloadSpeedOutput = "";
		String unit;
		Long bytesPrevious = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes();
		try {
			Thread.sleep(1000);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		Long bytesCurrent = TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes();
		long networkSpeed = bytesCurrent - bytesPrevious;
		if (networkSpeed >= GB) {
			unit = "GB/s";
		} else if (networkSpeed >= MB) {
			unit = "MB/s";
		} else {
			unit = "KB/s";
		}
		downloadSpeedOutput = String.valueOf(networkSpeed);
		String[] result = new String[] { downloadSpeedOutput, unit };
		return result;
	}

	public static long convertToBytes(float value, String unit) {
		if (unit == "GB/s") {
			return (long) value * GB;
		} else if (unit == "MB/s") {
			return (long) value * MB;
		} else if (unit == "KB/s") {
			return (long) value * KB;
		}
		return 0;
	}

	public static String getMetricData(long bytes) {
		float dataWithDecimal;
		String units;
		if (bytes >= GB) {
			dataWithDecimal = (float) bytes / GB;
			units = " GB";
		} else if (bytes >= MB) {
			dataWithDecimal = (float) bytes / MB;
			units = " MB";
		} else {
			dataWithDecimal = (float) bytes / KB;
			units = " KB";
		}
		return String.format("%.1f", dataWithDecimal) + units;
	}
}

File: app/src/main/java/otang/app/network/util/TrafficUtils.kt
package otang.app.network.util

import android.net.TrafficStats

object TrafficUtils {
    private const val GB: Long = 1000000000
    private const val MB: Long = 1000000
    private const val KB: Long = 1000
    val downloadSpeed: String
        get() {
            val speed: String
            val unit: String
            val bytesPrevious = TrafficStats.getTotalRxBytes()
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            val bytesCurrent = TrafficStats.getTotalRxBytes()
            val networkSpeed = bytesCurrent - bytesPrevious
            val downSpeed: Float
            if (networkSpeed >= GB) {
                downSpeed = (networkSpeed / GB).toFloat()
                unit = "GB/s"
            } else if (networkSpeed >= MB) {
                downSpeed = (networkSpeed / MB).toFloat()
                unit = "MB/s"
            } else {
                downSpeed = (networkSpeed / KB).toFloat()
                unit = "KB/s"
            }
            speed = String.format("%.1f", downSpeed)
            return speed + unit
        }
    val uploadSpeed: String
        get() {
            val speed: String
            val unit: String
            val bytesPrevious = TrafficStats.getTotalTxBytes()
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            val bytesCurrent = TrafficStats.getTotalTxBytes()
            val networkSpeed = bytesCurrent - bytesPrevious
            val upSpeed: Float
            if (networkSpeed >= GB) {
                upSpeed = (networkSpeed / GB).toFloat()
                unit = "GB/s"
            } else if (networkSpeed >= MB) {
                upSpeed = (networkSpeed / MB).toFloat()
                unit = "MB/s"
            } else {
                upSpeed = (networkSpeed / KB).toFloat()
                unit = "KB/s"
            }
            speed = String.format("%.1f", upSpeed)
            return speed + unit
        }
    val networkSpeed: Array<String>
        get() {
            val downloadSpeedOutput: String
            val unit: String
            val bytesPrevious =
                TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes()
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            val bytesCurrent =
                TrafficStats.getTotalRxBytes() + TrafficStats.getTotalTxBytes()
            val networkSpeed = bytesCurrent - bytesPrevious
            unit = if (networkSpeed >= GB) {
                "GB/s"
            } else if (networkSpeed >= MB) {
                "MB/s"
            } else {
                "KB/s"
            }
            downloadSpeedOutput = networkSpeed.toString()
            return arrayOf(downloadSpeedOutput, unit)
        }

    fun convertToBytes(value: Float, unit: String): Long {
        if (unit === "GB/s") {
            return value.toLong() * GB
        } else if (unit === "MB/s") {
            return value.toLong() * MB
        } else if (unit === "KB/s") {
            return value.toLong() * KB
        }
        return 0
    }

    fun getMetricData(bytes: Long): String {
        val dataWithDecimal: Float
        val units: String
        if (bytes >= GB) {
            dataWithDecimal = bytes.toFloat() / GB
            units = " GB"
        } else if (bytes >= MB) {
            dataWithDecimal = bytes.toFloat() / MB
            units = " MB"
        } else {
            dataWithDecimal = bytes.toFloat() / KB
            units = " KB"
        }
        return String.format("%.1f", dataWithDecimal) + units
    }
}